core/src/main/java/org/testcontainers/images/LoggedPullImageResultCallback.java

```diff
@@ -0,0 +1,112 @@
+package org.testcontainers.images;
+
+import com.github.dockerjava.api.model.PullResponseItem;
+import com.github.dockerjava.core.command.PullImageResultCallback;
+import org.slf4j.Logger;
+
+import java.io.Closeable;
+import java.util.*;
+
+import static java.lang.String.format;
+import static org.apache.commons.io.FileUtils.byteCountToDisplaySize;
+
+/**
+ * {@link PullImageResultCallback} with improved logging of pull progress.
+ */
+class LoggedPullImageResultCallback extends PullImageResultCallback {
+    private final Logger logger;
+
+    private Set<String> allLayers = new HashSet<>();
+    private Set<String> downloadedLayers = new HashSet<>();
+    private Set<String> pulledLayers = new HashSet<>();
+    private Map<String, Long> totalSizes = new HashMap<>();
+    private Map<String, Long> currentSizes = new HashMap<>();
+    private boolean completed;
+
+    LoggedPullImageResultCallback(final Logger logger) {
```

Comment:
`@RequiredArgsConstructor`, but also ok to keep it explicit

Reply:
I can plausibly foresee contributing this up to docker-java at some point, so would err on the side of keeping Lombok out particularly as it's very little difference either way.